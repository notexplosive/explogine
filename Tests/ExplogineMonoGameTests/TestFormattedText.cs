using System.Linq;
using ExplogineMonoGame.Data;
using FluentAssertions;
using Microsoft.Xna.Framework;
using Xunit;

namespace ExplogineMonoGameTests;

public class TestFormattedText
{
    [Fact]
    public void chars_in_equals_glyphs_out()
    {
        void TestCase(string input)
        {
            var formattedText = new FormattedText(new TestFont(), input);
            var glyphs = formattedText
                .GetGlyphs(new RectangleF(0, 0, float.MaxValue, float.MaxValue), Alignment.TopLeft).ToList();
            glyphs.Count.Should().Be(input.Length, $"\"{input}\" should have {input.Length} glyphs");
        }

        TestCase("There is just one line");
        TestCase("There\nis\none\nline\nper\nword");
        TestCase("There\n\nare\n\ntwo\n\nlines\n\nper\n\nword");
    }

    [Fact]
    public void one_newline_does_right_thing()
    {
        var font = new TestFont();
        var formattedText = new FormattedText(font, "String with\nOne newline");
        var glyphs = formattedText
            .GetGlyphs(new RectangleF(0, 0, float.MaxValue, float.MaxValue), Alignment.TopLeft).ToList();

        // this is a janky way to find the newline but at time of writing this test case the glyph count does not equal char count so there's really no good way to do this.
        foreach (var glyph in glyphs)
        {
            if (glyph.Data is FormattedText.FragmentChar fragmentChar)
            {
                if (fragmentChar.Text == 'O') // the O in "One newline"
                {
                    glyph.Position.Y.Should().Be(font.Height);
                }
            }
        }
    }

    [Fact]
    public void two_newlines_does_right_thing()
    {
        var font = new TestFont();
        var formattedText = new FormattedText(font, "String with\n\nTwo newline");
        var glyphs = formattedText
            .GetGlyphs(new RectangleF(0, 0, float.MaxValue, float.MaxValue), Alignment.TopLeft).ToList();

        foreach (var glyph in glyphs)
        {
            if (glyph.Data is FormattedText.FragmentChar fragmentChar)
            {
                if (fragmentChar.Text == 'T')
                {
                    glyph.Position.Y.Should().Be(font.Height * 2);
                }
            }
        }
    }
}

/// <summary>
///     Acts as a mock font for tests.
///     Maybe we can convert this to VirtualFont and bring it into product code?
/// </summary>
public class TestFont : IFont
{
    public IFont GetFont()
    {
        return this;
    }

    public float ScaleFactor => 1f;
    public float Height => 32;

    public Vector2 MeasureString(string text, float? restrictedWidth = null)
    {
        if (!restrictedWidth.HasValue)
        {
            return new Vector2(text.Length * ScaleFactor, Height);
        }

        return RestrictedStringBuilder.FromText(text, restrictedWidth.Value, this).Size;
    }
}
