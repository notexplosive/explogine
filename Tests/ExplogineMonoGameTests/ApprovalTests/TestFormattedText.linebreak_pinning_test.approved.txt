FormattedGlyph { Position = {X:0 Y:0}, Data = {X:32 Y:32} 'T' }
FormattedGlyph { Position = {X:32 Y:0}, Data = {X:32 Y:32} 'h' }
FormattedGlyph { Position = {X:64 Y:0}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:96 Y:0}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:128 Y:0}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:160 Y:0}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:192 Y:0}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:224 Y:0}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:256 Y:0}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:288 Y:0}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:320 Y:0}, Data = {X:32 Y:32} 'v' }
FormattedGlyph { Position = {X:352 Y:0}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:384 Y:0}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:416 Y:0}, Data = {X:32 Y:32} 'y' }
FormattedGlyph { Position = {X:448 Y:0}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:32}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:32 Y:32}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:64 Y:32}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:96 Y:32}, Data = {X:32 Y:32} 'g' }
FormattedGlyph { Position = {X:128 Y:32}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:160 Y:32}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:192 Y:32}, Data = {X:32 Y:32} 't' }
FormattedGlyph { Position = {X:224 Y:32}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:256 Y:32}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:288 Y:32}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:320 Y:32}, Data = {X:32 Y:32} 'g' }
FormattedGlyph { Position = {X:352 Y:32}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:384 Y:32}, Data = {X:32 Y:32} 'w' }
FormattedGlyph { Position = {X:416 Y:32}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:448 Y:32}, Data = {X:32 Y:32} 't' }
FormattedGlyph { Position = {X:480 Y:32}, Data = {X:32 Y:32} 'h' }
FormattedGlyph { Position = {X:0 Y:64}, Data = {X:0 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:64}, Data = {X:32 Y:32} 'm' }
FormattedGlyph { Position = {X:32 Y:64}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:64 Y:64}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:96 Y:64}, Data = {X:32 Y:32} 'y' }
FormattedGlyph { Position = {X:0 Y:96}, Data = {X:0 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:128}, Data = {X:0 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:128}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:32 Y:128}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:64 Y:128}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:96 Y:128}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:128 Y:128}, Data = {X:32 Y:32} 'b' }
FormattedGlyph { Position = {X:160 Y:128}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:192 Y:128}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:224 Y:128}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:256 Y:128}, Data = {X:32 Y:32} 'k' }
FormattedGlyph { Position = {X:288 Y:128}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:320 Y:128}, Data = {X:32 Y:32} '.' }
FormattedGlyph { Position = {X:352 Y:128}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:160}, Data = {X:32 Y:32} 'S' }
FormattedGlyph { Position = {X:32 Y:160}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:64 Y:160}, Data = {X:32 Y:32} 'm' }
FormattedGlyph { Position = {X:96 Y:160}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:128 Y:160}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:160 Y:160}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:192 Y:160}, Data = {X:32 Y:32} 'f' }
FormattedGlyph { Position = {X:224 Y:160}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:256 Y:160}, Data = {X:32 Y:32} 't' }
FormattedGlyph { Position = {X:288 Y:160}, Data = {X:32 Y:32} 'h' }
FormattedGlyph { Position = {X:320 Y:160}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:352 Y:160}, Data = {X:32 Y:32} 'm' }
FormattedGlyph { Position = {X:384 Y:160}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:192}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:32 Y:192}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:64 Y:192}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:96 Y:192}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:128 Y:192}, Data = {X:32 Y:32} 'm' }
FormattedGlyph { Position = {X:160 Y:192}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:192 Y:192}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:224 Y:192}, Data = {X:32 Y:32} 'u' }
FormattedGlyph { Position = {X:256 Y:192}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:288 Y:192}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:320 Y:192}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:224}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:32 Y:224}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:64 Y:224}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:96 Y:224}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:128 Y:224}, Data = {X:32 Y:32} 'b' }
FormattedGlyph { Position = {X:160 Y:224}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:192 Y:224}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:224 Y:224}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:256 Y:224}, Data = {X:32 Y:32} 'k' }
FormattedGlyph { Position = {X:288 Y:224}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:320 Y:224}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:352 Y:224}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:384 Y:224}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:416 Y:224}, Data = {X:32 Y:32} 'd' }
FormattedGlyph { Position = {X:448 Y:224}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:256}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:32 Y:256}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:64 Y:256}, Data = {X:32 Y:32} 'm' }
FormattedGlyph { Position = {X:96 Y:256}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:128 Y:256}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:160 Y:256}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:192 Y:256}, Data = {X:32 Y:32} 'f' }
FormattedGlyph { Position = {X:224 Y:256}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:256 Y:256}, Data = {X:32 Y:32} 't' }
FormattedGlyph { Position = {X:288 Y:256}, Data = {X:32 Y:32} 'h' }
FormattedGlyph { Position = {X:320 Y:256}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:352 Y:256}, Data = {X:32 Y:32} 'm' }
FormattedGlyph { Position = {X:384 Y:256}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:288}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:32 Y:288}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:64 Y:288}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:96 Y:288}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:128 Y:288}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:160 Y:288}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:192 Y:288}, Data = {X:32 Y:32} 't' }
FormattedGlyph { Position = {X:224 Y:288}, Data = {X:32 Y:32} 'u' }
FormattedGlyph { Position = {X:256 Y:288}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:288 Y:288}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:320 Y:288}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:352 Y:288}, Data = {X:32 Y:32} '.' }
FormattedGlyph { Position = {X:0 Y:320}, Data = {X:0 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:320}, Data = {X:32 Y:32} 'T' }
FormattedGlyph { Position = {X:32 Y:320}, Data = {X:32 Y:32} 'h' }
FormattedGlyph { Position = {X:64 Y:320}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:96 Y:320}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:128 Y:320}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:160 Y:320}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:192 Y:320}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:224 Y:320}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:256 Y:320}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:288 Y:320}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:320 Y:320}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:352 Y:320}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:384 Y:320}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:416 Y:320}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:448 Y:320}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:352}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:32 Y:352}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:64 Y:352}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:96 Y:352}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:128 Y:352}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:160 Y:352}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:192 Y:352}, Data = {X:32 Y:32} 'v' }
FormattedGlyph { Position = {X:224 Y:352}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:256 Y:352}, Data = {X:32 Y:32} 'r' }
FormattedGlyph { Position = {X:288 Y:352}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:320 Y:352}, Data = {X:32 Y:32} 'l' }
FormattedGlyph { Position = {X:352 Y:352}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:384}, Data = {X:32 Y:32} 'c' }
FormattedGlyph { Position = {X:32 Y:384}, Data = {X:32 Y:32} 'o' }
FormattedGlyph { Position = {X:64 Y:384}, Data = {X:32 Y:32} 'n' }
FormattedGlyph { Position = {X:96 Y:384}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:128 Y:384}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:160 Y:384}, Data = {X:32 Y:32} 'c' }
FormattedGlyph { Position = {X:192 Y:384}, Data = {X:32 Y:32} 'u' }
FormattedGlyph { Position = {X:224 Y:384}, Data = {X:32 Y:32} 't' }
FormattedGlyph { Position = {X:256 Y:384}, Data = {X:32 Y:32} 'i' }
FormattedGlyph { Position = {X:288 Y:384}, Data = {X:32 Y:32} 'v' }
FormattedGlyph { Position = {X:320 Y:384}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:352 Y:384}, Data = {X:32 Y:32} (whitespace) }
FormattedGlyph { Position = {X:0 Y:416}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:32 Y:416}, Data = {X:32 Y:32} 'p' }
FormattedGlyph { Position = {X:64 Y:416}, Data = {X:32 Y:32} 'a' }
FormattedGlyph { Position = {X:96 Y:416}, Data = {X:32 Y:32} 'c' }
FormattedGlyph { Position = {X:128 Y:416}, Data = {X:32 Y:32} 'e' }
FormattedGlyph { Position = {X:160 Y:416}, Data = {X:32 Y:32} 's' }
FormattedGlyph { Position = {X:192 Y:416}, Data = {X:32 Y:32} '.' }
